connect/api/src/main/java/org/apache/kafka/connect/sink/SinkRecord.java

```diff
@@ -57,6 +73,105 @@ public TimestampType timestampType() {
+
+    /**
+     * Get the original topic partition for this sink record, corresponding to the topic partition of the Kafka record
+     * before any {@link org.apache.kafka.connect.transforms.Transformation Transformation}s were applied. This should
+     * be used by sink tasks for any internal offset tracking purposes (that are reported to the framework via
+     * {@link SinkTask#preCommit(Map)} for instance) rather than {@link #kafkaPartition()}, in order to be compatible
+     * with transformations that mutate the topic partition value.
+     * <p>
+     * This method was added in Apache Kafka 3.6. Sink connectors that use this method but want to maintain backward
+     * compatibility in order to be able to be deployed on older Connect runtimes should guard the call to this method
+     * with a try-catch block, since calling this method will result in a {@link NoSuchMethodException} or
+     * {@link NoClassDefFoundError} when the sink connector is deployed to Connect runtimes older than Kafka 3.6.
+     * For example:
+     * <pre>{@code
+     * String originalKafkaPartition;
+     * try {
+     *     originalKafkaPartition = record.originalKafkaPartition();
+     * } catch (NoSuchMethodError | NoClassDefFoundError e) {
+     *     originalKafkaPartition = record.kafkaPartition();
+     * }
+     * }
+     * </pre>
+     * <p>
+     * Note that sink connectors that do their own offset tracking will be incompatible with SMTs that mutate the topic
+     * partition when deployed to older Connect runtimes.
+     *
+     * @return the topic partition corresponding to the Kafka record before any transformations were applied
+     *
+     * @since 3.6
+     */
+    public Integer originalKafkaPartition() {
+        return originalKafkaPartition;
+    }
+
+    /**
+     * Get the original offset for this sink record, corresponding to the offset of the Kafka record before any
+     * {@link org.apache.kafka.connect.transforms.Transformation Transformation}s were applied. This should be used by
+     * sink tasks for any internal offset tracking purposes (that are reported to the framework via
+     * {@link SinkTask#preCommit(Map)} for instance) rather than {@link #kafkaOffset()}, in order to be
+     * compatible with transformations that mutate the offset value.
+     * <p>
+     * This method was added in Apache Kafka 3.6. Sink connectors that use this method but want to maintain backward
+     * compatibility in order to be able to be deployed on older Connect runtimes should guard the call to this method
+     * with a try-catch block, since calling this method will result in a {@link NoSuchMethodException} or
+     * {@link NoClassDefFoundError} when the sink connector is deployed to Connect runtimes older than Kafka 3.6.
+     * For example:
+     * <pre>{@code
+     * String originalKafkaOffset;
+     * try {
+     *     originalKafkaOffset = record.originalKafkaOffset();
+     * } catch (NoSuchMethodError | NoClassDefFoundError e) {
```

Comment:
I think the NoClassDefFoundError was specific to the error reporter, since the method returned a type which didn't always exist. Here, where the return types are native Java types, I think the NoSuchMethodError is the only error possible here. Can you verify this?